File: src/main/java/net/snipsniper/secrets/games/BGame.java
package net.snipsniper.secrets.games;

import net.snipsniper.ImageManager;
import net.snipsniper.LogManager;
import net.snipsniper.SnipSniper;
import net.snipsniper.StatsManager;
import net.snipsniper.systray.Sniper;
import net.snipsniper.utils.enums.LogLevel;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.Random;

public class BGame extends JFrame {
    public final int INITIAL_SCREEN_WIDTH = 1024;
    public final int INITIAL_SCREEN_HEIGHT = 512;

    public final int BOARD_WIDTH = 10;
    public final int BOARD_HEIGHT = 20;

    private final int FALLSPEED_MAX_START = 500;

    private final int[] SCORES = {40, 100, 300, 1200};

    private final int LINES_BEFORE_LVLUP_ADD = 10;

    private BGameBlock[][] board;
    private final BGamePanel gamePanel = new BGamePanel(this);
    private BGamePiece cPiece;

    private boolean running = true;
    private boolean isPaused = false;

    private boolean[] keys;
    private BGameResources resources;

    private int score;
    private int level;
    private int rowsDone;
    private int rowsBeforeLevelUp = 10;

    private BGamePiece nextPiece;
    private boolean gameOver = false;
    private final Sniper sniper;

    private boolean hitDuringDownPress = false;

    public BGame(Sniper sniper) {
        this.sniper = sniper;
        StatsManager.incrementCount(StatsManager.BGAME_STARTED_AMOUNT);
        SnipSniper.getNewThread(args -> launch()).start();
    }

    private void launch() {
        resources = new BGameResources();
        resources.init();
        setTitle("Block Game");
        add(gamePanel);
        setIconImage(ImageManager.getImage("icons/random/kiwi.png"));
        addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent keyEvent) {
                super.keyPressed(keyEvent);
                keys[keyEvent.getKeyCode()] = true;

                if(keyEvent.getKeyCode() == KeyEvent.VK_ESCAPE && !gameOver)
                    isPaused = !isPaused;
            }

            @Override
            public void keyReleased(KeyEvent keyEvent) {
                super.keyReleased(keyEvent);
                keys[keyEvent.getKeyCode()] = false;
            }
        });
        addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                super.mouseClicked(e);
                if(e.getButton() == 3) {
                    gamePanel.screenshot();
                }
            }
        });
        addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                running = false;
                dispose();
            }
        });
        setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
        setVisible(true);
        gamePanel.setPreferredSize(new Dimension(INITIAL_SCREEN_WIDTH, INITIAL_SCREEN_HEIGHT));
        pack();
        Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
        int width = (int) screenSize.getWidth();
        int height = (int) screenSize.getHeight();
        setLocation(width / 2 - getWidth() / 2, height / 2 - getHeight() / 2);
        start();
        loop();
    }

    int fallSpeed = 0;
    int fallSpeedMax = FALLSPEED_MAX_START;
    public void loop() {
        while(running) {
            final int ts = getTileSize();
            setMinimumSize(new Dimension(BOARD_WIDTH *ts, BOARD_HEIGHT * ts));
            if(!isPaused) {
                input();
                if(!gameOver) {
                    boolean isHit = false;
                    if(cPiece != null)
                        isHit = cPiece.update();

                    if(fallSpeed >= fallSpeedMax) {
                        if(isHit || hitDuringDownPress) {
                            cPiece.hit();
                            hitDuringDownPress = false;
                            spawnPiece();
                        }
                        if(cPiece != null)
                            cPiece.moveDown();
                        fallSpeed = 0;
                    } else {
                        fallSpeed += 10;
                    }

                    int rows = checkRows();
                    if(rows != 0) {
                        rowsDone += rows;
                        if(rowsDone >= rowsBeforeLevelUp) {
                            rowsBeforeLevelUp += LINES_BEFORE_LVLUP_ADD;
                            level += 1;
                            fallSpeedMax -= 25;
                        }
                        int scoreMultiplier = level + 1;
                        score += SCORES[rows - 1] * scoreMultiplier;
                    }
                }
            }
            gamePanel.repaint();
            try {
                Thread.sleep(10);
            } catch (InterruptedException e) {
                LogManager.log("Error sleeping thread for BGame!", LogLevel.ERROR);
                LogManager.logStacktrace(e, LogLevel.ERROR);
            }
        }
    }

    public void gameOver() {
        gameOver = true;
    }

    public int checkRows() {
        int rowsCleared = 0;
        for(int y = 0; y < BOARD_HEIGHT; y++) {
            boolean hasFull = true;
            for(int x = 0; x < BOARD_WIDTH; x++) {
                if (board[x][y] == null) {
                    hasFull = false;
                    break;
                }
            }
            if(hasFull) {
                rowsCleared++;
                for(int x = 0; x < BOARD_WIDTH; x++)
                    board[x][y] = null;

                for(int z = y; z > 0; z--)
                    for(int x = 0; x < BOARD_WIDTH; x++)
                        board[x][z] = board[x][z-1];
            }
        }
        return rowsCleared;
    }

    public static int randomRange(int min, int max) {
        return new Random().nextInt((max - min) + 1) + min;
    }

    public void spawnPiece() {
        StatsManager.incrementCount(StatsManager.BGAME_STARTED_SPAWNED_PIECES_AMOUNT);
        BGamePiece newPiece = new BGamePiece(this);
        if(gameOver) {
            cPiece = null;
            return;
        }
        if(nextPiece == null)
            cPiece = newPiece;
        else
            cPiece = nextPiece;

        nextPiece = new BGamePiece(this);
    }

    public int getTileSize() {
        return gamePanel.getHeight() / BOARD_HEIGHT;
    }

    public void start() {
        gameOver = false;
        board = new BGameBlock[BOARD_WIDTH][BOARD_HEIGHT];
        fallSpeedMax = FALLSPEED_MAX_START;
        nextPiece = null;
        keys = new boolean[25565];
        score = 0;
        rowsDone = 0;
        level = 0;
        spawnPiece();
    }

    int dropCooldown = 0;
    int dropCooldownMax = 25;
    public void input() {
        if(dropCooldown > 0)
            dropCooldown--;

        if(isPressed(KeyEvent.VK_R)) {
            start();
            return;
        }

        if(cPiece != null) {
            if(isPressedAny(KeyEvent.VK_SPACE, KeyEvent.VK_SHIFT) && dropCooldown == 0 && !gameOver) {
                for(int i = 0; i < BOARD_HEIGHT; i++) {
                    if(cPiece.moveDown())
                        break;
                }
                dropCooldown = dropCooldownMax;
            }

            if(isPressedAny(KeyEvent.VK_E, KeyEvent.VK_UP) && !gameOver)
                cPiece.rotate(1);
            if(isPressed(KeyEvent.VK_Q) && !gameOver)
                cPiece.rotate(-1);

            if(isPressedAny(KeyEvent.VK_A, KeyEvent.VK_LEFT) && !gameOver)
                cPiece.move(-1);
            if(isPressedAny(KeyEvent.VK_D, KeyEvent.VK_RIGHT) && !gameOver)
                cPiece.move(1);

            if(isPressedAny(KeyEvent.VK_S, KeyEvent.VK_DOWN) && !gameOver) {
                cPiece.moveDown();
                if(cPiece.checkCollision())
                    hitDuringDownPress = true;
            }
        }
    }

    public boolean isPressed(int keyCode) {
        return keys[keyCode];
    }

    public boolean isPressedAny(int... keyCodes) {
        for(int keyCode : keyCodes) {
            if(keys[keyCode])
                return true;
        }
        return false;
    }

    public BGameResources getResources() {
        return resources;
    }

    public boolean isPaused() {
        return isPaused;
    }

    public BGamePiece getCurrentPiece() {
        return cPiece;
    }

    public BGameBlock[][] getBoard() {
        return board;
    }

    public int getScore() {
        return score;
    }

    public int getLevel() {
        return level;
    }

    public int getLinesCleared() {
        return rowsDone;
    }

    public boolean isGameOver() {
        return gameOver;
    }

    public BGamePiece getNextPiece() {
        return nextPiece;
    }

    public Sniper getSniper() {
        return sniper;
    }
}

File: src/main/kotlin/net/snipsniper/secrets/games/BGame.kt
package net.snipsniper.secrets.games

import net.snipsniper.SnipSniper.Companion.getNewThread
import net.snipsniper.StatsManager
import net.snipsniper.StatsManager.Companion.incrementCount
import net.snipsniper.systray.Sniper
import net.snipsniper.utils.getImage
import org.capturecoop.cclogger.CCLogLevel
import org.capturecoop.cclogger.CCLogger
import java.awt.Dimension
import java.awt.Toolkit
import java.awt.event.*
import java.util.*
import javax.swing.JFrame

class BGame(val sniper: Sniper) : JFrame() {
    lateinit var board: Array<Array<BGameBlock?>>
        private set
    private val gamePanel = BGamePanel(this)
    var currentPiece: BGamePiece? = null
        private set
    private var running = true
    var isPaused = false
        private set
    private var keys = Array(25565) { false }
    var resources: BGameResources? = null
        private set
    var score = 0
        private set
    var level = 0
        private set
    var linesCleared = 0
        private set
    private var rowsBeforeLevelUp = 10
    var nextPiece: BGamePiece? = null
        private set
    var isGameOver = false
        private set
    private var hitDuringDownPress = false
    private fun launch() {
        resources = BGameResources()
        resources!!.init()
        title = "Block Game"
        add(gamePanel)
        iconImage = "icons/random/block.png".getImage()
        addKeyListener(object : KeyAdapter() {
            override fun keyPressed(keyEvent: KeyEvent) {
                keys[keyEvent.keyCode] = true
                if (keyEvent.keyCode == KeyEvent.VK_ESCAPE && !isGameOver) isPaused = !isPaused
            }
            override fun keyReleased(keyEvent: KeyEvent) {
                keys[keyEvent.keyCode] = false
            }
        })
        addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(event: MouseEvent) = kotlin.run { if (event.button == 3) gamePanel.screenshot() }
        })
        addWindowListener(object : WindowAdapter() {
            override fun windowClosing(event: WindowEvent) {
                running = false
                dispose()
            }
        })
        defaultCloseOperation = DO_NOTHING_ON_CLOSE
        isVisible = true
        gamePanel.preferredSize = Dimension(INITIAL_SCREEN_WIDTH, INITIAL_SCREEN_HEIGHT)
        pack()
        Toolkit.getDefaultToolkit().screenSize.also {
            setLocation(it.width / 2 - width / 2, it.height / 2 - height / 2)
        }
        start()
        loop()
    }

    private var fallSpeed = 0
    private var fallSpeedMax = FALLSPEED_MAX_START
    private fun loop() {
        while (running) {
            val ts = tileSize
            minimumSize = Dimension(BOARD_WIDTH * ts, BOARD_HEIGHT * ts)
            if (!isPaused) {
                input()
                if (!isGameOver) {
                    var isHit = false
                    if (currentPiece != null) isHit = currentPiece!!.update()
                    if (fallSpeed >= fallSpeedMax) {
                        if (isHit || hitDuringDownPress) {
                            currentPiece!!.hit()
                            hitDuringDownPress = false
                            spawnPiece()
                        }
                        if (currentPiece != null) currentPiece!!.moveDown()
                        fallSpeed = 0
                    } else {
                        fallSpeed += 10
                    }
                    val rows = checkRows()
                    if (rows != 0) {
                        linesCleared += rows
                        if (linesCleared >= rowsBeforeLevelUp) {
                            rowsBeforeLevelUp += LINES_BEFORE_LVLUP_ADD
                            level += 1
                            fallSpeedMax -= 25
                        }
                        val scoreMultiplier = level + 1
                        score += SCORES[rows - 1] * scoreMultiplier
                    }
                }
            }
            gamePanel.repaint()
            try {
                Thread.sleep(10)
            } catch (e: InterruptedException) {
                CCLogger.log("Error sleeping thread for BGame!", CCLogLevel.ERROR)
                CCLogger.logStacktrace(e, CCLogLevel.ERROR)
            }
        }
    }

    fun gameOver() = kotlin.run { isGameOver = true }

    private fun checkRows(): Int {
        var rowsCleared = 0
        for (y in 0 until BOARD_HEIGHT) {
            var hasFull = true
            for (x in 0 until BOARD_WIDTH) {
                if (board[x][y] == null) {
                    hasFull = false
                    break
                }
            }
            if (hasFull) {
                rowsCleared++
                for (x in 0 until BOARD_WIDTH) board[x][y] = null
                for (z in y downTo 1) for (x in 0 until BOARD_WIDTH) board[x][z] = board[x][z - 1]
            }
        }
        return rowsCleared
    }

    private fun spawnPiece() {
        incrementCount(StatsManager.BGAME_STARTED_SPAWNED_PIECES_AMOUNT)
        if (isGameOver) {
            currentPiece = null
            return
        }
        currentPiece = nextPiece ?: BGamePiece(this)
        nextPiece = BGamePiece(this)
    }

    val tileSize: Int
        get() = gamePanel.height / BOARD_HEIGHT

    fun start() {
        board = Array(BOARD_WIDTH) { Array(BOARD_HEIGHT) { null } }
        isGameOver = false
        fallSpeedMax = FALLSPEED_MAX_START
        nextPiece = null
        score = 0
        linesCleared = 0
        level = 0
        spawnPiece()
    }

    private var dropCooldown = 0
    private val dropCooldownMax = 25

    init {
        incrementCount(StatsManager.BGAME_STARTED_AMOUNT)
        getNewThread { launch() }.start()
    }

    fun input() {
        if (dropCooldown > 0) dropCooldown--
        if (isPressed(KeyEvent.VK_R)) {
            start()
            return
        }
        if (currentPiece != null) {
            if (isPressedAny(KeyEvent.VK_SPACE, KeyEvent.VK_SHIFT) && dropCooldown == 0 && !isGameOver) {
                for (i in 0 until BOARD_HEIGHT) {
                    if (currentPiece!!.moveDown()) break
                }
                dropCooldown = dropCooldownMax
            }
            if (isPressedAny(KeyEvent.VK_E, KeyEvent.VK_UP) && !isGameOver) currentPiece!!.rotate(1)
            if (isPressed(KeyEvent.VK_Q) && !isGameOver) currentPiece!!.rotate(-1)
            if (isPressedAny(KeyEvent.VK_A, KeyEvent.VK_LEFT) && !isGameOver) currentPiece!!.move(-1)
            if (isPressedAny(KeyEvent.VK_D, KeyEvent.VK_RIGHT) && !isGameOver) currentPiece!!.move(1)
            if (isPressedAny(KeyEvent.VK_S, KeyEvent.VK_DOWN) && !isGameOver) {
                currentPiece!!.moveDown()
                if (currentPiece!!.checkCollision()) hitDuringDownPress = true
            }
        }
    }

    private fun isPressed(keyCode: Int) = keys[keyCode]

    private fun isPressedAny(vararg keyCodes: Int) = keyCodes.any { keys[it] }

    companion object {
        private const val INITIAL_SCREEN_WIDTH = 1024
        private const val INITIAL_SCREEN_HEIGHT = 512
        const val BOARD_WIDTH = 10
        const val BOARD_HEIGHT = 20
        private const val FALLSPEED_MAX_START = 500
        private val SCORES = intArrayOf(40, 100, 300, 1200)
        private const val LINES_BEFORE_LVLUP_ADD = 10

        fun randomRange(min: Int, max: Int) = Random().nextInt(max - min + 1) + min
    }
}